feat: mark user editable regions in help post

KravMaguy opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #46411 

<!-- Feel free to add any additional description of changes below this line -->

Ive tested this by achieving the desired output as posted in the issue:

![image](https://user-images.githubusercontent.com/37171082/201772101-0e04be75-1e10-44ab-93c3-63fcec15a890.png)

Commit:
mark editable regions in help post

KravMaguy commented:
This PR is not ready

KravMaguy commented:
Not ready to be merged, does not work if adding multiple lines or deleting multiple lines, also if editing the challenge again after posting.

Commit:
case where lines were added or removed

naomi-lgbt commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -4,6 +4,7 @@ import { mapTo, tap } from 'rxjs/operators';
 
 import envData from '../../../../../config/env.json';
 import { transformEditorLink } from '../utils';
+import { SuperBlocks } from '../../../../../config/certification-settings';
```
```suggestion
```

If my other suggestion is accepted, this import will be unnecessary.

naomi-lgbt reviewed (changes requested):
This is a great start! A couple of thoughts:

jeremylt commented:
I'm not familiar enough with the internals to comment on implementation, but the output looks right.

naomi-lgbt commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -50,6 +51,34 @@ function createQuestionEpic(action$, state$, { window }) {
+      if (superBlock === SuperBlocks.RespWebDesignNew) {
+        const editableRegionStrings = new (function () {
+          this.editableRegionStr = 'User Editable Region';
+          this.startComment = '<!---';
+          this.endComment = '--->';
```
Correct.

KravMaguy replied:
completed, now we can work on the second part of the requirement, show diff outside of the editable region.

jeremylt replied:
Note - that **should not** be part of this PR

KravMaguy committed:
Update client/src/templates/Challenges/redux/create-question-epic.js

allow regions to be marked in any file that has editable boundaries

Co-authored-by: Naomi Carrigan <nhcarrigan@gmail.com>

KravMaguy committed:
Update client/src/templates/Challenges/redux/create-question-epic.js

remove constant, it is no longer necessary since condition is checking for editable boundaries.

Co-authored-by: Naomi Carrigan <nhcarrigan@gmail.com>

Commit:
update comments per appropriate file extension

Commit:
fix typeError cannot read properties of undefined

KravMaguy commented:
Reguarding your suggestion of using the editableBoundaries instead of the superblock constants, (when the issue was originally posted, I had been under the impression that you only wanted this feature for the resp design curriculum). 

At one point the app crashed and I encountered a TypeError. I've added the appropriate fix (I hope). 

![Capture_challenge](https://user-images.githubusercontent.com/37171082/202529032-b90ad7a7-97fa-4663-834b-9aa80075aac7.JPG)

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -50,6 +50,41 @@ function createQuestionEpic(action$, state$, { window }) {
+        challengeFiles?.some(file => file.editableRegionBoundaries.length > 0)
+      ) {
+        const editableRegionStrings = fileExtension => {
+          const lineBreak = '\n';
+          const multiLineStart = '/*';
+          const multiLineEnd = '*/';
+          const htmlStart = '<!---';
+          const htmlEnd = '--->';
+          const editableRegionStr = 'User Editable Region';
+          const startComment =
+            fileExtension === 'html' ? htmlStart : multiLineStart;
+          const endComment = fileExtension === 'html' ? htmlEnd : multiLineEnd;
+          const editableRegionComments = {
+            comment: `${lineBreak}${startComment} ${editableRegionStr} ${endComment}${lineBreak}`
+          };
+          return { lineBreak, editableRegionComments };
```
```suggestion
          const startComment = fileExtension === 'html' ? '<!--' : '/*';
          const endComment = fileExtension === 'html' ? '-->' : '*/';
          return `\n${startComment} 'User Editable Region' ${endComment}\n`;
```
A more concise version like this should be easier to read.  Also, the html comments were `<!---` instead of `<!--`, so I changed that.

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -50,6 +50,41 @@ function createQuestionEpic(action$, state$, { window }) {
+        challengeFiles.forEach(file => {
+          const { contents, editableRegionBoundaries, ext } = file;
+          if (editableRegionBoundaries.length > 0) {
+            const editableRegionStr = editableRegionStrings(ext);
+            const {
+              lineBreak,
+              editableRegionComments: { comment }
+            } = editableRegionStr;
```
```suggestion
            const comment = editableRegionStrings(ext);
```
The nesting seemed redundant.

ojeytonwilliams reviewed (changes requested):
The functionality looks good, but I think we can refactor things a little.

KravMaguy committed:
Update client/src/templates/Challenges/redux/create-question-epic.js

suggested refactor constants

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

KravMaguy committed:
Update client/src/templates/Challenges/redux/create-question-epic.js

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

KravMaguy committed:
Update client/src/templates/Challenges/redux/create-question-epic.js

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -50,6 +50,27 @@ function createQuestionEpic(action$, state$, { window }) {
+            const editableContents = contents.split('\n');
+            editableContents.splice(start, 0, comment);
+            editableContents.splice(end, 0, comment);
+            file.contents = editableContents.join('\n');
```
Sorry, I didn't spot this before, but we can't mutate properties of `challengeFiles`.  If we do that it mutates the redux store.  To see the impact

1. Click the ask for help button on, say, http://localhost:8000/learn/2022/responsive-web-design/learn-html-by-building-a-cat-photo-app/step-2
2. Hide the editor by clicking on `index.html`
3. Re-open the editor and see

![image](https://user-images.githubusercontent.com/15801806/204343743-b8fc7b06-b728-472f-a07f-aab4a640f14a.png)

The quick fix is to re-write this to use a clone of `challengeFiles`, but the nicer fix is to avoid mutation entirely.

KravMaguy replied:
Thank you for demonstrating the problem so clearly. Ive fixed it accordingly.

Commit:
avoid mutating challengeFiles

Commit:
refactor for efficiency

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -43,13 +43,37 @@ function createQuestionEpic(action$, state$, { window }) {
+        };
+
+        const filesWithEditableRegions = challengeFiles.map(file => {
+          const { contents: Contents, editableRegionBoundaries, ext } = file;
```
```suggestion
          const { contents, editableRegionBoundaries, ext } = file;
```
This isn't necessary, we can just use `contents` as is.  Also, we typically use camelCase for variables and PascalCase for classes.

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -43,13 +43,37 @@ function createQuestionEpic(action$, state$, { window }) {
+          if (editableRegionBoundaries.length > 0) {
+            const comment = editableRegionStrings(ext);
+            const [start, end] = editableRegionBoundaries;
+            const editableContents = Contents.split('\n');
+            editableContents.splice(start, 0, comment);
+            editableContents.splice(end, 0, comment);
+            const contents = editableContents.join('\n');
```
```suggestion
            const lines = contents.split('\n');
            lines.splice(start, 0, comment);
            lines.splice(end, 0, comment);
```
It's a little misleading, since `editableContents` has a specific meaning in the codebase - literally the part of the contents that is inside the editable region.

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -43,13 +43,37 @@ function createQuestionEpic(action$, state$, { window }) {
+            editableContents.splice(start, 0, comment);
+            editableContents.splice(end, 0, comment);
+            const contents = editableContents.join('\n');
+            return { ...file, contents };
```
```suggestion
            return { ...file, contents: lines.join('\n') };
```
The property name, 'contents', is enough of an explanation imo.

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -43,13 +43,37 @@ function createQuestionEpic(action$, state$, { window }) {
+            const contents = editableContents.join('\n');
+            return { ...file, contents };
+          }
+          return { ...file };
```
```suggestion
          return file;
```
As long as we don't mutate, it's not necessary to defensively copy objects.

ojeytonwilliams reviewed (changes requested):
Hey, @KravMaguy, thanks for taking care of that bug.

Aside from some minor style issues, I think this is good to go!

Commit:
suggested style changes

KravMaguy commented:
here it says one of the end to end tests were not successful
1) JavaScript projects can be submitted and then viewed in /settings and on the certifications 

ive run the tests in this cypress file locally and they have all passed:

![image](https://user-images.githubusercontent.com/37171082/206894176-63dbb3d9-c7df-4608-b140-07049e537e12.png)

naomi-lgbt commented:
The cypress tests can be flaky sometimes - I've triggered a re-run on the failed test.

ojeytonwilliams commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -43,13 +43,36 @@ function createQuestionEpic(action$, state$, { window }) {
+            lines.splice(end, 0, comment);
+            return { ...file, contents: lines.join('\n') };
+          }
+          return { file };
```
```suggestion
          return file;
```
Otherwise we're nesting `file` too deeply.

KravMaguy committed:
Update client/src/templates/Challenges/redux/create-question-epic.js

return file exactly as is if no editable boundaries.

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams reviewed (approved):
LGTM :+1:

Thanks, @KravMaguy!

naomi-lgbt commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -43,13 +43,36 @@ function createQuestionEpic(action$, state$, { window }) {
+        const editableRegionStrings = fileExtension => {
+          const startComment = fileExtension === 'html' ? '<!--' : '/*';
+          const endComment = fileExtension === 'html' ? '-->' : '*/';
+          return `\n${startComment} 'User Editable Region' ${endComment}\n`;
```
```suggestion
          return `\n${startComment} User Editable Region ${endComment}\n`;
```

I'm nitpicking - but we don't need quotes in the comment it looks weird.

raisedadead commented:
We are officially on Node.js 18 now. We have changed the merge requirements accordingly.

You will need to rebase this PR against the main branch OR Click the "Update branch" button to get the new checks going.

Thanks for your contribution.

raisedadead committed:
Update client/src/templates/Challenges/redux/create-question-epic.js

Co-authored-by: Naomi Carrigan <nhcarrigan@gmail.com>
